### ISSUE TITLE
Request: Adding a way to randomise key items and spell levels

### ISSUE THREAD

**Issue by AvatarMew** [gameboy9/FFRPSP#3]
This would enable to game to be even more random.
If the above is possible is there a way to add a logic like in the NES randomiser so that the game is beatable?
Also if it is possible to add the key items to the pool of items that can be found in an chest is it possible to prevent them from being found in any of the optional dungeons?
It would allow races to occur.

**Comment by gameboy9** [comment]
I would LOVE to randomize key items.  The problem is that it also sets a trigger and I can't seem to move such a box around for the life of me.  But if I figure it out at some point, and I think that I can, I would totally do it.

I also would LOVE to scramble the spell levels around a la NES Randomizer, but, unfortunately, when I thought I was successful, it screwed up the monster patterns, which is why I came up with the solution that I did; basically randomizing the prices and MP costs in a way where Level 8 Flare could operate like a level 1 spell.  (50G/3MP cost)  

I also randomized who can learn what spells as follows: 
75% chance for each spell for each of the white/black mages, 
100% for the white/black wizards, naturally.  
50% chance for red mages for all spells
An additional 50% chance for red wizards of learning a spell the Red Mage didn't learn.  
33% chance for Ninjas.

**Comment by AvatarMew** [comment]
So, level one flare is currently impossible? That's quite sad. I hope that it'll one day be possible!
When the spell and randomised their mp is also randomised?

It would be great if you could find a way to add the key items in the pool of items!

A major issue with randomising key items is that there would be a chance they would be placed inside the optional dungeons. It would make the game almost impossible to beat..
. I don't know how that could be avoided.

**Comment by gameboy9** [comment]
No, I can't, for some reason, declare Flare to be level 1.  I can only give it costs that act like a level 1 spell and basically give all mages level 8 access. (but they can't have certain spells as explained before)  So it is possible to cast Flare from experience level 1 for 50G and 3MP per cast.

The key items issue is a big issue, surely.  Fortunately, however, they basically section off the bonus dungeons as a separate piece of code, so it's not a nasty obstacle to avoid.

**Comment by gameboy9** [comment]
To be clear, I can tell any treasure chest to contain a key item, and it is awarded, but there's a trigger that isn't set, and that's what the program is looking for when trying to advance the plot.

**Comment by AvatarMew** [comment]
Ah, okay. I understand..
Thank you for the replies. 

Is it possible to have a slider that affects the monsters; stats like the NES randomiser has?
Would there be a "remove permissions" option to enable you to use higher magic spells?

**Comment by gameboy9** [comment]
Oh, randomized monster stats?  Yes, I thought that I had implemented it, but apparently I didn't.  I'll take care of that for sure.  I might implement an option to randomize experience and gold rewards as well.  I also want to implement a randomize weapon/armor powers (power, accuracy, and evasion) track as well.  (up to 20-500%)

**Comment by AvatarMew** [comment]
I'll look forward to these additions!

One more thing, the program doesn't work with a EUR iso, it appears to only work with a US one.

**Comment by gameboy9** [comment]
Yes, that is most likely true, and will probably remain that way for the foreseeable future.

**Comment by AvatarMew** [comment]
I know this is a long shot but when/if you're able to randomise the key items do you think it'll be possible to have an option to get the canoe/go to Ordeals early? 
Oh, I think that the Levistone and the Adamatite can't be shuffled as they're sprites.

**Comment by gameboy9** [comment]
I'm not sure if early Canoe is going to be possible, but I'll keep trying!

I thought I had a breakthrough with the key items, but I, unfortunately, ran into another major obstacle that I have to jump over...

**Comment by AvatarMew** [comment]
What was the obstacle?

**Comment by gameboy9** [comment]
When I tried to change the trigger from, say, retrieving the crown to retrieving the TNT, the game softlocks.  :(

**Comment by AvatarMew** [comment]
Oh, that's quite worrying. At least it's a softlock and not a hardlock...

**Comment by gameboy9** [comment]
So here's what's going down:
- Each location has a "text pointer" that points to the text of each event, including picking up key items, and possibly including what NPCs say.
- This is why I can't simply change the crown to the TNT because they are in two different locations and the text points to two completely different locations.
- I *THINK* that it is relatively easy for me to change the text around, but it will be VERY tedious!
- Therefore, I think I have found a way to randomize key items.
- HOWEVER, this is going to take a REALLY long time.

Here's what is really easy to do:  Early Canoe.  Or early ship.  Or early TNT.  Or early Lich defeat.  Or early TOFR.  Or early... whatever you can think of!  This is because every plot event is trigger-based.(which is also why randomized key items is hard to accomplish)  I will have to figure out all of the triggers, but when I do, this shouldn't be too hard to take care of.

**Comment by AvatarMew** [comment]
Ah, I understand why it would be hard to do it. I',m happy you're going to attempt it!

I'm glad that easy access to the canoe can be done.

**Comment by AvatarMew** [comment]
Are there any updates?

**Comment by AvatarMew** [comment]
Hello, I know you've not worked on this program for a LONG time but what I've suggested has been done on the GBA version of FFI so maybe you could ask them for help?

**Comment by nic0lette** [comment]
I'm just going to capture this information for people who may want to dig into this some more :)

The scripted events use a custom scripting bytecode that has commands for checking plot flags (talked to the king, lit the earth crystal, etc...), giving/taking key items (jolt tonic, etc...) These scripts are (from what I can tell at a first glance) the same as the GBA version, and so the documentation on the event commands by Kea/Vennobennu in [Final Fantasy I: Dawn of Souls General Notes](https://www.romhacking.net/documents/750/) should be accurate.

For example, the initalization code for the town of Corneria is

GBA:
```
2d 08 28 02 00 7a 00 08    # Check to see if you've talked to the King, if you haven't jump to the end
2e 04 8c 13                # Remove one of the guards at the exits
2e 04 8c 13                # Remove one of the guards at the exits
2e 04 8c 13                # Remove one of the guards at the exits
00 04 ff ff                # End
```

PSP:
```
2d 08 28 02 64 3c 01 08    # Check to see if you've talked to the King, if you haven't jump to the end
2e 04 8c 13                # Remove one of the guards at the exits
2e 04 8c 13                # Remove one of the guards at the exits
2e 04 8c 13                # Remove one of the guards at the exits
2d 08 2d 02 74 3c 01 08.   # Check new flag for the PSP version
...
```

I think the notes from the GBA version (and the code from HMS Jayne) could be useful references for digging into this more.

As for spell levels, I don't know the details, but do know that there's a binary patch that's applied to the GBA ROM so this works. The notes I have for it say:

```
-A spell's level is determined by its data byte, not by ID:
	Field Spell Menu 1 - print all spell names:
		$3983E: C907 1BD0 517A 0000
		$39850: 8842 12DB
	Field Spell Menu 2 - trying to cast current spell ($2FC34):
		507A 317F 8142 06DA 78E2
	Shopping ($47078):
		0248 1101 4118 497A 08E0 0000 80191A08
	Shopping 2 - Generic determination? ($454FA):
		0348 1901 4018 407A 0138 06E0 0000 80191A08
	Adding spell to your spellbook ($44F7A):
		875C 3B01 024A D318 587A 441E 21E0 80191A08
	($451D8):
		C75C 014C 3901 15E0 80191A08
	($4520C):
		6118 4C7A 013C
```

Hopefully this might be a hint to someone who wants to get that part to work as well. :)
